File: packages/design-ir/src/history.ts
import type { DesignDocument, OperationExecution } from "./types";

export interface HistoryEntry {
  readonly operation_ids: readonly string[];
  readonly before: DesignDocument;
  readonly after: DesignDocument;
}

export class CommandHistory {
  #entries: HistoryEntry[] = [];
  #cursor = 0;

  push(before: DesignDocument, execution: OperationExecution): void {
    this.#entries = this.#entries.slice(0, this.#cursor);
    this.#entries.push({
      operation_ids: [...execution.applied_operation_ids],
      before: structuredClone(before),
      after: structuredClone(execution.document),
    });
    this.#cursor = this.#entries.length;
  }

  undo(current: DesignDocument): DesignDocument {
    if (this.#cursor === 0) return structuredClone(current);
    this.#cursor -= 1;
    return structuredClone(this.#entries[this.#cursor]!.before);
  }

  redo(current: DesignDocument): DesignDocument {
    if (this.#cursor >= this.#entries.length) return structuredClone(current);
    const result = structuredClone(this.#entries[this.#cursor]!.after);
    this.#cursor += 1;
    return result;
  }

  get size(): number {
    return this.#entries.length;
  }

  get cursor(): number {
    return this.#cursor;
  }
}

File: packages-py/design_ir/history.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass

from .models import DesignDocument, OperationExecution


@dataclass(frozen=True, slots=True)
class HistoryEntry:
    operation_ids: tuple[str, ...]
    before: DesignDocument
    after: DesignDocument


class CommandHistory:
    def __init__(self) -> None:
        self._entries: list[HistoryEntry] = []
        self._cursor = 0

    def push(self, before: DesignDocument, execution: OperationExecution) -> None:
        self._entries = self._entries[: self._cursor]
        self._entries.append(
            HistoryEntry(
                execution.applied_operation_ids,
                deepcopy(before),
                deepcopy(execution.document),
            )
        )
        self._cursor = len(self._entries)

    def undo(self, current: DesignDocument) -> DesignDocument:
        if self._cursor == 0:
            return deepcopy(current)
        self._cursor -= 1
        return deepcopy(self._entries[self._cursor].before)

    def redo(self, current: DesignDocument) -> DesignDocument:
        if self._cursor >= len(self._entries):
            return deepcopy(current)
        result = deepcopy(self._entries[self._cursor].after)
        self._cursor += 1
        return result

    @property
    def size(self) -> int:
        return len(self._entries)

    @property
    def cursor(self) -> int:
        return self._cursor
